server/src/test/java/org/elasticsearch/index/IndexModuleTests.java

```diff
@@ -718,6 +730,79 @@ public void onIndexCommitDelete(ShardId shardId, IndexCommit deletedCommit) {
         }
     }
 
+    public void testCustomRecoverySourceFactory() throws IOException {
+        IndexModule module = new IndexModule(
+            indexSettings,
+            emptyAnalysisRegistry,
+            InternalEngine::new,
+            Collections.emptyMap(),
+            () -> true,
+            indexNameExpressionResolver,
+            Collections.emptyMap()
+        );
+
+        class CustomReplicationTracker extends ReplicationTracker {
+            CustomReplicationTracker(
+                ShardId shardId,
+                String allocationId,
+                IndexSettings indexSettings,
+                long operationPrimaryTerm,
+                long globalCheckpoint,
+                LongConsumer onGlobalCheckpointUpdated,
+                LongSupplier currentTimeMillisSupplier,
+                BiConsumer<RetentionLeases, ActionListener<ReplicationResponse>> onSyncRetentionLeases,
+                Supplier<SafeCommitInfo> safeCommitInfoSupplier,
+                Consumer<ReplicationGroup> onReplicationGroupUpdated
+            ) {
+                super(
+                    shardId,
+                    allocationId,
+                    indexSettings,
+                    operationPrimaryTerm,
+                    globalCheckpoint,
+                    onGlobalCheckpointUpdated,
+                    currentTimeMillisSupplier,
+                    onSyncRetentionLeases,
+                    safeCommitInfoSupplier,
+                    onReplicationGroupUpdated
+                );
+            }
+        }
+        module.setReplicationTrackerFactory(
```

Comment:
Can you update `testFrozen()` as well to ensure it cannot be set once the IndexModule is frozen?